Asserting different effects (#503)
lightbend/akkaserverless-java-sdk: docs/src/modules/java/pages/actions.adoc

include::example$scala-fibonacci-action/src/main/scala/com/example/fibonacci/Main.scala[]
----

By default, the generated constructor has an `ActionCreationContext` parameter, but you can change this to accept other parameters. 
If you change the constructor of the `FibonacciAction` class you will see a compilation error here, and you have to adjust the 
factory function that is passed to `AkkaServerlessFactory.withComponents`.

When more components are added the `AkkaServerlessFactory` is regenerated, and you have to adjust the registration from the `Main` class.

== Testing the Action 

=== Unit tests

The following snippet shows how the `CounterJournalToTopicActionTestKit` is used to test the `CounterJournalToTopicAction` implementation.

AkkaServerless provides the `CounterJournalToTopicActionTestKit` that allows us to call the methods of `CounterJournalToTopicAction`. For each `Action` AkkaServerless generates an specific test kit for it, with the name `[ActionName]TestKit`. This test kit hold an instance of its `Action`. Each call we execute over to the test kit returns an `ActionResult` that holds the effect produced by the underlying `CounterJournalToTopicAction`. 

> Apart from the test kit, to facilitate the creation of unit tests, AkkaServerless generates test classes, ready to use, based on the Action defined in the .proto files. This is shown in the snippet below. 

Reviewer: ```suggestion
Apart from the test kit AkkaServerless generates test classes based on the Action defined in the .proto files. This is shown in the snippet below. 
```

We can have a more straightforward text if we remove the `, something in the middle,`. In that case here, it's clear that we generate it to facilitate the creation of tests and that the generated tests is ready for use.

Is worth noting that Actions are unique units of computation. The framework doesn't reuse an Action instance. A consequence of this is that we should create a test kit for each method we evaluate as shown in the snippet. 

.src/test/java/com/example/actions/CounterJournalToTopicActionTest.java
[source,java]
----
include::example$java-eventsourced-counter/src/test/java/com/example/actions/CounterJournalToTopicActionTest.java[tag=class] 
----
<1> creating the test kit to test `increase` Action's method.
<2> calling `increase` method with some value.
<3> retrieving the `Reply` from `increase` method.

*ActionResult*

> Calling an entity method through the TestKit gives us back an link:{attachmentsdir}/testkit/com/akkaserverless/javasdk/testkit/ActionResult.html[`ActionResult` {tab-icon}, window="new"]. This class has methods that we can use to assert our expectations, such as:

Reviewer: ```suggestion
Calling an action method through the TestKit gives us back an link:{attachmentsdir}/testkit/com/akkaserverless/javasdk/testkit/ActionResult.html[`ActionResult` {tab-icon}, window="new"]. This class has methods that we can use to assert our expectations, such as:
```

* `getReply` returns the value passed to the `Effect.reply`, if any.
* `getForwardServiceCall` returns the `ServiceCall` to use to forward the call, if any.

> *CounterJournalToTopicActionTestKit*

Reviewer 2: I don't think you need to mention the generated test kit again, it's introduced in the text above.

Author: less is more

> * `getReply` is the response of the method called in the test
> * `getForwardServiceCall` returns the `ServiceCall` to which is fowarded the command the `Action` receives.

Reviewer: this two sentences are not so clear to me, what about...

```suggestion
* `getReply` returns the value passed to the `Effect.reply`, if any.
* `getForwardServiceCall` returns the `ServiceCall` to use to forward the call, if any.
```

The following snippet shows how the `CounterJournalToTopicActionTestKit` is used to test the `CounterJournalToTopicAction` implementation.

> AkkaServerless generates the `CounterJournalToTopicActionTestKit` that allows us to call the methods of `CounterJournalToTopicAction`. For each `Action` Akka Serverless generates a specific test kit for it, with the name `[ActionName]TestKit`. This test kit hold an instance of its `Action`. Each call we execute over to the test kit returns an `ActionResult` that holds the effect produced by the underlying `CounterJournalToTopicAction`. 

Reviewer 2: And one more in the paragraph below
